tests: add initial test suite for `CoreMenu`

EshaanAgg opened the pull request:
## Summary
- Adds the complete tests for `CoreMenuDivider`
- The `CoreMenuOption` component has two major parts:
   - Toggle component which controls where the subroutes are to be rendered or not
   - The subroutes
In this PR, I add the complete test suite for the former, and plan to add the one for later in the next PR for easier reviews
- Added a validator for the `subRoutes` props based on the usage of the same in the component

## Testing checklist

- [X] Contributor has fully tested the PR manually
- [X] If there are any front-end changes, before/after screenshots are included
- [X] Critical user journeys are covered by Gherkin stories
- [X] Critical and brittle code paths are covered by unit tests


## PR process

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

EshaanAgg committed:
feat: init the test suite for the CoreMenu items

EshaanAgg committed:
feat: add more test cases for user interaction in CoreMenuOption

EshaanAgg committed:
add validator

AlexVelezLl commented on kolibri/core/assets/src/views/CoreMenu/__tests__/CoreMenuOption.spec.js:
```diff
@@ -0,0 +1,165 @@
+};
+
+describe('CoreMenuOption', () => {
+  describe('subRoutes toggle', () => {
```
Here we can have first a `smoke test` that just renders the component and check that it doesnt break [e.g.](https://github.com/EshaanAgg/kolibri/blob/4a3a4160420d8cb7b7fe39c96e455d122a5aa09f/kolibri/core/assets/src/views/sync/FacilityNameAndSyncStatus/__tests__/index.spec.js#L5)

EshaanAgg replied:
Will add the same!

AlexVelezLl commented on kolibri/core/assets/src/views/CoreMenu/__tests__/CoreMenuOption.spec.js:
```diff
@@ -0,0 +1,165 @@
+        expect(screen.getByTestId('icon-chevronUp')).toBeInTheDocument();
+        expect(screen.queryByTestId('icon-chevronDown')).not.toBeInTheDocument();
+
+        // Clicking again should hide the subroutes
```
This could be another test, to make sure we are testing just one thing.

AlexVelezLl commented on kolibri/core/assets/src/views/CoreMenu/CoreMenuOption.vue:
```diff
@@ -46,7 +46,7 @@
     </li>
 
-    <div v-if="visibleSubMenu">
+    <div v-if="visibleSubMenu" data-testid="subroutes-container">
```
It seems that this testid is not used

AlexVelezLl commented on kolibri/core/assets/src/views/CoreMenu/__tests__/CoreMenuOption.spec.js:
```diff
@@ -0,0 +1,165 @@
+import { render, fireEvent, screen } from '@testing-library/vue';
```
I think that now that we are doing tests for the CoreMenu, I think it would be good to do tests for the index as well, since it has interesting behavior with keyboard navigation for example.

EshaanAgg replied:
I tried adding a new test suite for checking the focusability like the following:
```ts
     it('pressing tab from keyboard should focus the menuitem', async () => {
        renderComponent({ subRoutes: sampleSubRoutes });

        // Press tab to focus the menuitem
        await fireEvent.keyDown(document.body, { key: 'Tab', code: 'Tab' });

        const menuItem = screen.getByRole('menuitem');
        expect(menuItem).toHaveFocus();
      });
```
but I can't seem to get it to work. I tried a couple of different combinations:
- Replacing `document.body` with `document`, `document.activeElement`, `menuItem` etc.
- Replacing the Key object for `fireEvent` with other variations like involving keys like `keyCode`
- Pressing the Tab key multiple times
However, this does not seem to shift the focus from the default `body` to any other element. Can you please help me figure out why the same doesn't work?

AlexVelezLl replied:
Oh for this we can use the [userEvent](https://testing-library.com/docs/user-event/v13/) package provided by testing-library! We will need to install it here. I thought that we had used it in these tests :sweat_smile:. With this you can use the `userEvent.tab` method:

```javascript
  it('pressing tab from keyboard should focus the menuitem', async () => {
    renderComponent({ subRoutes: sampleSubRoutes });

    // Press tab to focus the menuitem
    await userEvent.tab(); // skip csrfmiddlewaretoken input
    await userEvent.tab();

    const menuItem = screen.getByRole('menuitem');
    expect(menuItem).toHaveFocus();
  });
```

Note that I added an aditional userEvent.tab because for some reason we add a csrfmiddlewaretoken input [here](https://github.com/EshaanAgg/kolibri/blob/77048d3b922e0a1bd91731804ff07cb348dc28ae/packages/kolibri-tools/jest.conf/setup.js#L58). Let me check why do we have that, but for now, this is working :).

AlexVelezLl replied:
Just checked with Richard and told me that this csrfmiddlewaretoken input is not needed anymore, so feel free to remove it :).

EshaanAgg replied:
Thanks @AlexVelezLl for all the help! I had stumbled upon the `userEvent` package too, but wasn't able to decipher what it's doing differently than `fireEvent`, so I was skeptical about adopting it. But using `userEvent.tab()` indeed does work! Have updated the PR with the same, and even removed the `csrfmiddlewaretoken` input.

AlexVelezLl reviewed (commented):
Thanks @EshaanAgg! You are making great progress towards using vue-testing library in Kolibri!! I left just a couple of comments and considerations :).

EshaanAgg commented:
Hi! I'm so sorry. It seems like this the updates on the PR just completely got lost in my inbox. Would work on addressing the review. Thanks a lot @AlexVelezLl!

EshaanAgg committed:
address review

EshaanAgg committed:
add user event and fix tests

EshaanAgg committed:
merge upstream

EshaanAgg committed:
fix dependencies

EshaanAgg committed:
add rewire package

AlexVelezLl commented on packages/kolibri-tools/package.json:
```diff
@@ -67,6 +70,7 @@
     "query-ast": "^1.0.5",
     "readline-sync": "^1.4.9",
     "recast": "^0.23.6",
+    "rewire": "^7.0.0",
```
rewire before the change had a version of `^6.0.0` Is there any reason why we changed to 7.0.0? This could bring some unintended errors

EshaanAgg replied:
My bad. I'll revert the same.

AlexVelezLl commented on yarn.lock:
```diff
@@ -2,26 +2,24 @@
 # yarn lockfile v1
 
 
+"@aashutoshrathi/word-wrap@^1.2.3":
```
It seems that somehow the yarn.lock got messed, could you try to fix it? One way to solve it can be copying the yarn.lock that is already in `develop`, and run `yarn install` again, make sure that the diffs are only the ones corresponding to the `userEvent` library. After doing this. please feel free of making a git squash of the last 3 commits so that the blame does not point to you.

EshaanAgg replied:
Got it. I have resolved the issue with the `yarn.lock`. I tried squashing the last few commits with rebase, but I ran into a couple of merge conflicts, so I bailed on it. Do we still need the squashing given now the diff in `yarn.lock` is only limited to `userEvent`?

AlexVelezLl replied:
Yes, although the PR shows the diff only of the user event change, the blame of almost the whole yarn.lock is still on you because the change was made in a previous commit. But no problem, I can squash it in the PR merge.

I'll give it one last check, and it'll be ready to merge. Thank you! :open_hands:

EshaanAgg replied:
Ahh got it. I'll try to be more careful the next time! Thanks a lot for the guidance and help!

EshaanAgg committed:
revert yarn.lock

EshaanAgg committed:
update yarn.lock

EshaanAgg committed:
merge upstream

AlexVelezLl commented on kolibri/core/assets/src/views/CoreMenu/__tests__/CoreMenuOption.spec.js:
```diff
@@ -0,0 +1,202 @@
+        const menuItem = screen.getByRole('menuitem');
+
+        // Clicking should show the subroutes
+        await fireEvent.click(menuItem);
```
Im sorry I havent noticed this before, until now I thought we were using the user-event library :sweat_smile:, but I think Its worth getting used to preferring the userEvent library instead of fireEvent, as it is the [recommended way](https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#not-using-testing-libraryuser-event) for the most use cases, so we can replace all these clicks and keydowns methods of fireEvent with the userEvent.click or userEvent.type methods.

EshaanAgg replied:
Hey! Nice catch. I have updated all the instances of `fireEvent.click` with `userEvent.click` as instructed. The `userEvent` library does not provide a method to simulate a `keyDown` event with respect to a particular element, as we require in the case of `menuitem` and the `Enter` key. I tried using the `.keyboard` and `.type` methods, but they didn't seem to trigger the desired effects. Is there any way that I'm missing?

AlexVelezLl replied:
(Just for the reference) It seems that with `userEvent.type` was not working because before typing on the element, it internally clicks on the element to have the focus. And as we have the handler also on the `click` element, it was calling the function twice, so visibleSubMenu=!visibleSubMenu called twice ends in the sub menu not being visible. So its ok to keep the fireEvent here.

EshaanAgg replied:
Ahh got it! Thanks so much for the explanation!

AlexVelezLl commented on kolibri/core/assets/src/views/CoreMenu/__tests__/CoreMenuOption.spec.js:
```diff
@@ -0,0 +1,202 @@
+import userEvent from '@testing-library/user-event';
+import VueRouter from 'vue-router';
+import CoreMenuOption from '../CoreMenuOption.vue';
+import '@testing-library/jest-dom';
```
I think we can now remove this import, as we already have it in the jest `setup.js` file.

EshaanAgg replied:
I tried that, but it didn't work. Removing the import for `userEvent` started giving a linting error, as well as an error when running the test suite (saying that `userEvent` is not defined).

AlexVelezLl replied:
Oh I was talking about the `import '@testing-library/jest-dom';` :sweat_smile: but it seems that you already deleted it :hugs:

AlexVelezLl reviewed (commented):
Thank you for all your patience! @EshaanAgg Promise that now these will be last comments about things that I had not noticed before. After this we should be ready to go =).

EshaanAgg committed:
make use of userEvent

AlexVelezLl reviewed (approved):
Thanks for your work here @EshaanAgg! LGTM!!
